uniffi_bindgen/src/bindings/swift/templates/Template-Bridging-Header.h

```diff
@@ -6,6 +6,14 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+// The following structs are used to implement the lowest level
+// of the FFI, and thus useful to multiple uniffied crates.
+// We ensure they are declared exactly once, with a header guard, UNIFFI_SHARED_H.
+// This may become a potential footgun if multiple versions of uniffi are being used
+// in the same project.
+#ifndef UNIFFI_SHARED_H
+#define UNIFFI_SHARED_H
```

Comment:
Note: foot-gun potential

Reply:
This is making me realize that I don't have a fully fleshed-out mental model of how the uniffi helper code works across versions, and it's probably worth us discussing a bit at a high-level to get aligned.

In the initial commits for the project, I deliberately generated all of the helper code alongside the module's own code, rather than e.g. making a "uniffi_runtime" Kotlin package that the generated code could depend on. This was partly for simplicity, and partly from a half-baked notion that you might be able to safely depend on multiple uniffied packages that were using multiple different versions of uniffi. But it's not clear whether we actually need that in practice, or what the costs of supporting that are going to be.

I'm fine with carefully moving away from that model, and assuming that all the uniffied components in a build are using the same version of uniffi, as long as we can mitigate the footgun potential.

What do you think about trying to put a little version check in here? Something like (untested!):

```
#ifdef UNIFFI_SHARED_H
  #ifndef UNIFFI_VERSION_1_2_3
  #error Combining helper code from multiple versions of uniffi is not supported
  #endif
#else
  #define UNIFFI_SHARED_H
  #define UNIFFI_VERSION_1_2_3
  // ...the generated header code
#endif
```

We'd have to do a bit of plumbing to expose the current version number in a suitable form, but that doesn't seem too bad.

Reply:
Your `(untested!)` scheme is stricter than I was thinking, but probably more sensible. 

I'll leave the scheme in my head here, only to get it out of my head.

```h
#ifndef UNIFFI_SHARED_0_7_2
#define UNIFFI_SHARED_0_7_2
// everthing defined up until now.
#endif
#ifndef UNIFFI_SHARED_0_8_0
#define UNIFFI_SHARED_0_8_0
// A theoretical non-breaking addition in 0.8.0
#endif
```

But I only have a partial solution for breaking changes.

It strikes me that this is very much bread and butter for more experienced C developers than I.

@dmose may be able to point to prior art here.

Reply:
I don't have much experience in the space of handling use of multiple SDK versions of the same thing in an executable at compile time.

That said, I'm a fan of the "not supported" approach, because I think it keeps us from expanding our problem space to include complexity that we don't currently have reason to believe we'll need.

If you really want more C/C++ thoughts, there are some C/C++ experts at Mozila (eg Botond) that one could reach out to.

Reply:
> // A theoretical non-breaking addition in 0.8.0

FWIW, I don't really trust us (including myself!) to reason carefully about whether or not each change is going to be sufficiently non-breaking, or even to have the tools to really understand the ways breaking things could happen. We could get there with deliberate effort, but I just don't think we're operating in that mode right now, or that it's a useful way to spend our limited energies at this time.

Rephrased: I think we should deliberately choose not to expend effort on layers of b/w compat right now, it's low value and high risk for the current state of the project. Like so many of our choices here, we should also deliberately plan to revisit that question in future.

Reply:
> I think we should deliberately choose not to expend effort on layers of b/w compat right now

:100: I agree.